R/response_process.R

```diff
@@ -0,0 +1,190 @@
+#' Process a Google API response
+#'
+#' @description
+#' Intended primarily for internal use in client packages that provide
+#' high-level wrappers for users. Typically applied as the final step in this
+#' sequence of calls:
+#'   * Request prepared with [request_build()].
+#'   * Request made with [request_make()].
+#'   * Response processed with [response_process()].
+#'
+#' All that's needed for a successful request is to parse the JSON extracted via
+#' `httr::content()`. Therefore, the main point of `response_process()` is to
+#' handle less happy outcomes:
+#'   * Status code in the 100s (information) or 300s (redirection). These are
+#'     unexpected.
+#'   * Non-JSON content type, such as HTML.
+#'   * Status codes in the 400s (client error) and 500s (server error). The
+#'     structure of the error payload varies across Google APIs and we try to
+#'     create a useful message for all variants we know about.
+
+#'
+#' @details
+#' A redacted version of the `resp` input is returned in the condition (auth
+#' tokens are removed). Use functions such as `rlang::last_error()` or
+#' `rlang::catch_cnd()` to capture the condition and do a more detailed forensic
+#' examination.
+#'
+#' @param resp Object of class `response` from [httr].
+#' @param raw Logical. Whether to return parsed content (`FALSE`) or in `raw`
```

Comment:
I actually took this feature directly from bigrquery and have never had it in the analogous functions in googledrive or googlesheets(4).

I thought about it and decided perhaps it would be useful for knocking out intermediate steps, sometimes?

A little research to inform our discussion:

I believe this is used (exactly?) once in bigrquery:

https://github.com/r-dbi/bigrquery/blob/b5b7ccf30a939ed57d722b807818197c11bb8faa/R/bq-download.R#L232

`bq_get()` is the only such request-wrapper that even exposes `raw`, so I suspect this is why the argument exists?

Reply:
If we remove the `raw` argument, it would be nice to refactor `response_process()` in such a way that the above bigrquery usage has another reasonable solution.